### PR TITLE
Use CHANMODES when modifying channel modes

### LINKED ISSUE #144: Certain MODE messages could access on undefined
Presently I'm looking at https://github.com/martynsmith/node-irc/blob/35ee099b0517f47c51577403973b7416d1d29201/lib/irc.js#L229-L239 which does not verify that "user" is actually a user present in the channel.

This has occasionally crashed when a non-user value is passed as an argument:

```
[01 06:44:12] *** Mode #defocus +q *!*@adsl-76-237-180-172.dsl.chcgil.sbcglobal.net by mrmist
 TypeError: Cannot call method 'indexOf' of undefined
     at irc/lib/irc.js:233:54
```

Usually it works fine though, and I can't quite figure out why. But examining the code, almost certainly the server is capable of crafting a response that will throw a TypeError, disconnecting the IRC connection, and this should never happen.

On a similar note, the debug should probably look more like so:
```javascript
                if ( self.opt.debug )
                    util.log("MODE:" + message.args[0] + " sets mode: " + message.args.slice(1).join(' '));
```

### PR DESCRIPTION
**Pull request by awwright**
This commit enhances how channel modes are set, and should be free of any TypeError errors. It also creates a new Function Channels since that's the most efficient way of creating multiple Objects with the same properties, and adds a method to set that channel's mode, emitting events as necessary if there is an associated message to emit with.

Closes #144.

### CODE REVIEW

**Comment by apeiron** [comment]
@Acubed Do you know why @expr closed this?

**Comment by Acubed** [comment]
@apeiron I didn't even know it was closed >:|

@expr is there any way I can help fix #222?

**Comment by apeiron** [comment]
@Acubed Additionally, how does this improve upon #222?

**Comment by Acubed** [comment]
@apeiron The server defines which modes are "channel" modes, they're not hard-coded into IRC. #222 would serve to fix some servers and break others.

File: lib/irc.js [hunk 1]
```diff
@@ -123,7 +123,6 @@ function Client(server, nick, opt) {
                                 value = value.split(',');
                                 var type = ['a','b','c','d']
                                 for (var i = 0; i < type.length; i++) {
-
```
Comment (apeiron) [inline lib/irc.js]: Nitpick: remove this empty line.

**Comment by apeiron** [comment]
@Acubed Can you update this so it will merge cleanly, and we can revisit it?

**Comment by Acubed** [comment]
New branch pulled in #270

**Comment by Acubed** [comment]
Reopen for master branch.

There's a craptonne more merge conflicts, I'll essentially have to re-write the code it looks like.

**Comment by Acubed** [comment]
I can't seem to reopen this. GitHub, halp

**Comment by apeiron** [comment]
That's odd. I get "Something went wrong with that request. Please try again."

**Comment by apeiron** [comment]
@Acubed I have spoken with Github support and they deployed a fix to allow me to re-open this pull request. You'll need to (rebase)[https://github.com/martynsmith/node-irc/wiki/0.3.x-branch-debacle] the pull request. Thank you!

**Comment by awwright** [comment]
Closing because it appears this was fixed in #351

**Pull request closed without merging**